// @blueprintjs/popover2/lib/cjs/classes.js
"use strict";
/* !
 * Copyright 2021 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOOLTIP2_INDICATOR = exports.TOOLTIP2 = exports.POPOVER2_TRANSITION_CONTAINER = exports.POPOVER2_TARGET = exports.POPOVER2_OPEN = exports.POPOVER2_DISMISS_OVERRIDE = exports.POPOVER2_DISMISS = exports.POPOVER2_CONTENT_SIZING = exports.POPOVER2_CONTENT_PLACEMENT = exports.POPOVER2_CONTENT = exports.POPOVER2_CAPTURING_DISMISS = exports.POPOVER2_BACKDROP = exports.POPOVER2_ARROW = exports.POPOVER2 = exports.CONTEXT_MENU2_BACKDROP = exports.CONTEXT_MENU2_POPOVER2 = exports.CONTEXT_MENU2_VIRTUAL_TARGET = exports.CONTEXT_MENU2 = void 0;
var core_1 = require("@blueprintjs/core");
var NS = core_1.Classes.getClassNamespace();
exports.CONTEXT_MENU2 = NS + "-context-menu2";
exports.CONTEXT_MENU2_VIRTUAL_TARGET = exports.CONTEXT_MENU2 + "-virtual-target";
exports.CONTEXT_MENU2_POPOVER2 = exports.CONTEXT_MENU2 + "-popover2";
exports.CONTEXT_MENU2_BACKDROP = exports.CONTEXT_MENU2 + "-backdrop";
exports.POPOVER2 = NS + "-popover2";
exports.POPOVER2_ARROW = exports.POPOVER2 + "-arrow";
exports.POPOVER2_BACKDROP = exports.POPOVER2 + "-backdrop";
exports.POPOVER2_CAPTURING_DISMISS = exports.POPOVER2 + "-capturing-dismiss";
exports.POPOVER2_CONTENT = exports.POPOVER2 + "-content";
exports.POPOVER2_CONTENT_PLACEMENT = exports.POPOVER2 + "-placement";
exports.POPOVER2_CONTENT_SIZING = exports.POPOVER2_CONTENT + "-sizing";
exports.POPOVER2_DISMISS = exports.POPOVER2 + "-dismiss";
exports.POPOVER2_DISMISS_OVERRIDE = exports.POPOVER2_DISMISS + "-override";
exports.POPOVER2_OPEN = exports.POPOVER2 + "-open";
exports.POPOVER2_TARGET = exports.POPOVER2 + "-target";
exports.POPOVER2_TRANSITION_CONTAINER = exports.POPOVER2 + "-transition-container";
exports.TOOLTIP2 = NS + "-tooltip2";
exports.TOOLTIP2_INDICATOR = exports.TOOLTIP2 + "-indicator";
